pandas/core/frame.py

```diff
@@ -4831,17 +4832,60 @@ def reindex(self, *args, **kwargs) -> DataFrame:
         kwargs.pop("labels", None)
         return super().reindex(**kwargs)
 
-    @deprecate_nonkeyword_arguments(version=None, allowed_args=["self", "labels"])
+    @overload
+    def drop(
+        self,
+        labels: Hashable | list[Hashable] = ...,
+        *,
+        axis: Axis = ...,
+        index: Hashable | list[Hashable] = ...,
+        columns: Hashable | list[Hashable] = ...,
+        level: Level | None = ...,
+        inplace: Literal[True],
+        errors: IgnoreRaise = ...,
+    ) -> None:
+        ...
+
+    @overload
     def drop(
         self,
-        labels=None,
+        labels: Hashable | list[Hashable] = ...,
+        *,
+        axis: Axis = ...,
+        index: Hashable | list[Hashable] = ...,
+        columns: Hashable | list[Hashable] = ...,
+        level: Level | None = ...,
+        inplace: Literal[False] = ...,
+        errors: IgnoreRaise = ...,
+    ) -> DataFrame:
+        ...
+
+    @overload
+    def drop(
+        self,
+        labels: Hashable | list[Hashable] = ...,
+        *,
+        axis: Axis = ...,
+        index: Hashable | list[Hashable] = ...,
+        columns: Hashable | list[Hashable] = ...,
+        level: Level | None = ...,
+        inplace: bool = ...,
+        errors: IgnoreRaise = ...,
+    ) -> DataFrame | None:
+        ...
+
+    # error: Signature of "drop" incompatible with supertype "NDFrame"
+    @deprecate_nonkeyword_arguments(version=None, allowed_args=["self", "labels"])
+    def drop(  # type: ignore[override]
+        self,
```

Comment:
I think we discussed this in another PR, but it's unclear to me why the signatures are incompatible.  Is it because you don't have `self: DataFrame` here?  Or the type of `inplace` is `bool` versus `bool_t` ?  Or something else?

Reply:
I think @simonjayhawkins pointed out that this is a result of the parent implementation not having keyword-only arguments, even though the parent overloads use keyword-only arguments. If that is the case, these errors will disappear after the deprecation phase.

`bool` vs `bool_t` has no impact.

Reply:
I was hoping that either the keyword-only issue or the return type (generic vs. the same class) are the cause for that but in this toy example, I do not need an ignore comment - I don't know why. (Probably missing something obvious)

```py
from __future__ import annotations

from typing import (
    Literal,
    TypeVar,
    overload,
)

ParentT = TypeVar("ParentT", bound="Parent")


class Parent:
    @overload
    def test(self, *, a: Literal[False]) -> None:
        ...

    @overload
    def test(self: ParentT, *, a: Literal[True] = ...) -> ParentT:
        ...

    @overload
    def test(self: ParentT, *, a: bool = ...) -> ParentT | None:
        ...

    def test(self: ParentT, a: bool = False) -> ParentT | None:
        ...


class Child(Parent):
    @overload
    def test(self, *, a: Literal[False]) -> None:
        ...

    @overload
    def test(self, *, a: Literal[True] = ...) -> Child:
        ...

    @overload
    def test(self, *, a: bool = ...) -> Child | None:
        ...

    def test(self, a: bool = False) -> Child | None:
        ...
```

Reply:
The decorator seems to irritate mypy, even though the decorator seems to be correctly typed.

Opened https://github.com/python/mypy/issues/12387

Reply:
Glad your figured it out that it is the decorator.  Can you add a comment in this PR near the `#type: ignore` line that points to the mypy issue?